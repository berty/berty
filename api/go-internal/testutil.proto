syntax = "proto3";

package testutil;

option go_package = "berty.tech/berty/go/internal/testutil";

service TestService {
  // use to test unary
  rpc EchoTest(EchoTest.Request) returns (EchoTest.Reply);

  // Use to test stream.
  rpc EchoStreamTest(EchoStreamTest.Request) returns (stream EchoStreamTest.Reply);

  // Use to test duplex stream.
  rpc EchoDuplexTest(stream EchoDuplexTest.Request) returns (stream EchoDuplexTest.Reply);
}

message EchoTest {
  message Request {
    int64 delay = 1;
    string echo = 2;
    bool trigger_error = 3;
  }
  message Reply {
    string echo = 1;
  }
}

message EchoStreamTest {
  message Request {
    int64 delay = 1;
    string echo = 2;
    bool trigger_error = 3;
  }
  message Reply {
    string echo = 1;
  }
}

message EchoDuplexTest {
  message Request {
    string echo = 1;
    bool trigger_error = 2;
  }
  message Reply {
    string echo = 1;
  }
}
