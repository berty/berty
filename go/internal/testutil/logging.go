package testutil

import (
	"flag"
	"os"
	"strings"
	"sync"
	"testing"

	"github.com/stretchr/testify/assert"
	"go.uber.org/zap"

	"berty.tech/berty/v2/go/internal/logutil"
)

const defaultLoggingFilters = "info+:bty.test* error+:*,-ipfs*"

var (
	logFilters     = flag.String("log-filters", defaultLoggingFilters, "log namespaces")
	logFile        = flag.String("log-file", "", "log to file")
	logFormat      = flag.String("log-format", "color", "json, console, color")
	loggerInstance *zap.Logger
	loggerCleanup  func()
	loggerInitOnce sync.Once
)

func Logger(t *testing.T) (*zap.Logger, func()) {
	t.Helper()

	loggerInitOnce.Do(func() {
		if val := os.Getenv("BERTY_LOGFILTERS"); val != "" {
			*logFilters = val
		}
		if val := os.Getenv("BERTY_LOGFILE"); val != "" {
			*logFile = val
		}
		if val := os.Getenv("BERTY_LOGFORMAT"); val != "" {
			*logFormat = val
		}
		*logFilters = strings.ReplaceAll(*logFilters, ":default:", defaultLoggingFilters)

		var err error
		loggerInstance, loggerCleanup, err = logutil.NewLogger(*logFilters, *logFormat, *logFile)
		if !assert.NoError(t, err) {
			loggerInstance = zap.NewNop()
			loggerCleanup = func() {}
		}
	})

	return loggerInstance, loggerCleanup
}
