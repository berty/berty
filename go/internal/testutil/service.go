package testutil

import (
	context "context"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"go.uber.org/zap"
	"google.golang.org/grpc"

	"berty.tech/berty/v2/go/pkg/errcode"
	"berty.tech/weshnet/v2/pkg/grpcutil"
)

type service struct {
	logger *zap.Logger

	UnimplementedTestServiceServer
}

type Options struct {
	Logger *zap.Logger
}

func (o *Options) applyDefault() {
	if o.Logger == nil {
		o.Logger = zap.NewNop()
	}
}

func TestingNewServiceClient(ctx context.Context, t *testing.T, opts *Options) (*grpc.ClientConn, *grpc.Server) {
	t.Helper()

	srv := grpc.NewServer()
	svc := NewService(opts)
	RegisterTestServiceServer(srv, svc)

	l := grpcutil.NewBufListener(2048)

	cc, err := l.NewClientConn(ctx)
	assert.NoError(t, err)

	go func() {
		_ = srv.Serve(l.Listener)
	}()

	t.Cleanup(func() { l.Close() })

	return cc, srv
}

func NewService(opts *Options) TestServiceServer {
	opts.applyDefault()
	return &service{
		logger: opts.Logger,
	}
}

func (svc *service) EchoTest(_ context.Context, req *EchoTest_Request) (*EchoTest_Reply, error) {
	if req.TriggerError {
		return nil, errcode.ErrCode_ErrTestEcho
	}

	if req.Delay > 0 {
		time.Sleep(time.Duration(req.Delay) * time.Millisecond)
	}

	return &EchoTest_Reply{Echo: req.Echo}, nil
}

func (svc *service) EchoStreamTest(req *EchoStreamTest_Request, srv TestService_EchoStreamTestServer) error {
	if req.TriggerError {
		return errcode.ErrCode_ErrTestEcho
	}

	for {
		err := srv.Send(&EchoStreamTest_Reply{Echo: req.Echo})
		if err != nil {
			return errcode.ErrCode_ErrTestEchoSend.Wrap(err)
		}

		time.Sleep(time.Duration(req.Delay) * time.Millisecond)
	}
}

func (svc *service) EchoDuplexTest(srv TestService_EchoDuplexTestServer) error {
	for {
		req, err := srv.Recv()
		if err != nil {
			return errcode.ErrCode_ErrTestEchoRecv.Wrap(err)
		}

		if req.TriggerError {
			return errcode.ErrCode_ErrTestEcho
		}

		err = srv.Send(&EchoDuplexTest_Reply{
			Echo: req.Echo,
		})
		if err != nil {
			return errcode.ErrCode_ErrTestEchoSend.Wrap(err)
		}
	}
}
