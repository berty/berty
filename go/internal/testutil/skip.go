package testutil

import (
	"os"
	"testing"
)

func SkipSlow(t *testing.T) {
	t.Helper()
	if os.Getenv("SKIP_SLOW_TESTS") == "1" {
		t.Skip("slow test skipped")
	}
}
