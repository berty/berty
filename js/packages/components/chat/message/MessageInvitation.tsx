import { CommonActions } from '@react-navigation/native'
import { Icon } from '@ui-kitten/components'
import { Buffer } from 'buffer'
import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import { ActivityIndicator, TouchableOpacity, View } from 'react-native'

import {
	HorizontalButtons,
	SecondaryButtonIconLeft,
	TertiaryButtonIconLeft,
} from '@berty/components/buttons'
import { useStyles } from '@berty/contexts/styles'
import { useOneToOneContact, useConversation } from '@berty/hooks'
import { useNavigation } from '@berty/navigation'
import { useMessengerClient, useThemeColor } from '@berty/store'
import { InteractionGroupInvitation } from '@berty/store/types.gen'
import { base64ToURLBase64 } from '@berty/utils/convert/base64'

import { MultiMemberAvatar } from '../../avatars'
import { UnifiedText } from '../../shared-components/UnifiedText'
import { MessageSystemWrapper } from './MessageSystemWrapper'

export const MessageInvitationButton: React.FC<{
	onPress?: () => void
	activeOpacity: any
	backgroundColor: any
	icon: any
	color: any
	title: string
	styleOpacity?: any
	disabled?: boolean
	loading?: boolean
}> = ({
	onPress,
	activeOpacity,
	backgroundColor,
	icon,
	color,
	title,
	styleOpacity,
	disabled = false,
	loading,
}) => {
	const { flex, padding, border, width, row, text, opacity } = useStyles()
	return (
		<TouchableOpacity
			style={[flex.align.center]}
			activeOpacity={activeOpacity}
			onPress={onPress}
			disabled={disabled}
		>
			<View
				style={[
					padding.tiny,
					padding.vertical.small,
					border.radius.tiny,
					width(120),
					row.center,
					flex.align.center,
					padding.right.small,
					{ backgroundColor },
				]}
			>
				{loading ? (
					<ActivityIndicator />
				) : (
					<Icon name={icon} width={24} height={24} fill={color} style={[opacity(styleOpacity)]} />
				)}
				<UnifiedText
					style={[
						text.align.center,
						text.bold,
						opacity(styleOpacity),
						{ color, textTransform: 'uppercase' },
					]}
				>
					{title}
				</UnifiedText>
			</View>
		</TouchableOpacity>
	)
}

const MessageInvitationSent: React.FC<{ message: InteractionGroupInvitation }> = ({ message }) => {
	const { text } = useStyles()
	const { t } = useTranslation()

	const conversationContact = useOneToOneContact(message.conversationPublicKey || '')
	return (
		<UnifiedText style={[text.size.scale(14), text.align.center]}>
			{`${t('chat.one-to-one.contact-request-box.you-invited')} ${
				conversationContact?.displayName || t('chat.one-to-one.contact-request-box.this-contact')
			} ${t('chat.one-to-one.contact-request-box.to-a-group')}`}
		</UnifiedText>
	)
}

const MessageInvitationReceived: React.FC<{ message: InteractionGroupInvitation }> = ({
	message,
}) => {
	const { row, flex, text, margin } = useStyles()
	const colors = useThemeColor()
	const client = useMessengerClient()
	const { t } = useTranslation()
	const { dispatch } = useNavigation()
	const [error, setError] = useState(false)
	const [{ convPk, displayName }, setPdlInfo] = useState({ convPk: '', displayName: '' })
	const [accepting, setAccepting] = useState(false)
	const conv = useConversation(convPk)
	const { link } = message.payload || {}
	const acceptDisabled = !!(accepting || conv || !convPk || error)

	// Parse deep link
	React.useEffect(() => {
		if (client && !convPk && link && !conv) {
			setError(false)
			client
				.parseDeepLink({
					link,
				})
				.then(reply => {
					setPdlInfo({
						displayName: reply.link?.bertyGroup?.displayName || '',
						convPk: base64ToURLBase64(
							Buffer.from(reply.link?.bertyGroup?.group?.publicKey || '').toString('base64'),
						),
					})
				})
				.catch(err => {
					console.warn(err)
					setError(true)
				})
		}
	}, [convPk, conv, link, client])

	const handleAccept = React.useCallback(() => {
		if (client && convPk && !conv && !accepting && !error) {
			setAccepting(true)
			client
				.conversationJoin({ link })
				.then(() => {
					dispatch(
						CommonActions.reset({
							routes: [
								{ name: 'Chat.Home' },
								{
									name: 'Chat.Group',
									params: {
										convId: convPk,
									},
								},
							],
						}),
					)
				})
				.catch(err => {
					console.warn(err)
					setError(true)
				})
				.finally(() => {
					setAccepting(false)
				})
		}
	}, [client, convPk, conv, accepting, error, link, dispatch])

	return (
		<>
			<View style={[row.left, flex.align.center, flex.justify.center]}>
				<UnifiedText style={[text.bold, { textTransform: 'uppercase' }]}>
					{t('chat.one-to-one.contact-request-box.group-invitation')}
				</UnifiedText>
			</View>
			<View style={[margin.top.small, flex.align.center, flex.justify.center]}>
				<View style={margin.bottom.small}>
					<MultiMemberAvatar publicKey={convPk} size={40} fallbackNameSeed={displayName} />
				</View>
				<UnifiedText style={[text.size.small, text.light, margin.bottom.small]}>
					{displayName}
				</UnifiedText>
			</View>
			<View style={[margin.top.medium, margin.horizontal.large]}>
				<HorizontalButtons>
					<TertiaryButtonIconLeft
						disabled
						name='close-outline'
						onPress={undefined} // TODO: Command to refuse invitation
					>
						{t('chat.one-to-one.contact-request-box.refuse-button')}
					</TertiaryButtonIconLeft>
					<SecondaryButtonIconLeft onPress={handleAccept} disabled={acceptDisabled ? true : false}>
						{!conv
							? t('chat.one-to-one.contact-request-box.accept-button')
							: t('chat.one-to-one.contact-request-box.accepted-button')}
					</SecondaryButtonIconLeft>
				</HorizontalButtons>
			</View>
			{error && (
				<UnifiedText
					style={[
						margin.top.small,
						margin.horizontal.small,
						text.size.small,
						text.light,
						text.align.center,
						{ color: colors['warning-asset'] },
					]}
				>
					{t('chat.one-to-one.contact-request-box.error')}
				</UnifiedText>
			)}
		</>
	)
}

export const MessageInvitation: React.FC<{ message: InteractionGroupInvitation }> = ({
	message,
}) => {
	const { row, padding, margin } = useStyles()

	return (
		<View style={[row.center, padding.horizontal.medium, margin.bottom.small, { paddingTop: 2 }]}>
			{message.isMine ? (
				<MessageSystemWrapper logo={false}>
					<MessageInvitationSent message={message} />
				</MessageSystemWrapper>
			) : (
				<MessageSystemWrapper>
					<MessageInvitationReceived message={message} />
				</MessageSystemWrapper>
			)}
		</View>
	)
}
