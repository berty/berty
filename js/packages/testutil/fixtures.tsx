import beapi from '@berty-tech/api'

const empty: beapi.messenger.IMedia = {
	metadataBytes: beapi.messenger.MediaMetadata.encode({}).finish(),
}

const media = {
	empty,
}

export default {
	media,
}
