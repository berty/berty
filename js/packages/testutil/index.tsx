import React from 'react'
import { NavigationContainer } from '@react-navigation/native'
import { IconRegistry } from '@ui-kitten/components'
import { EvaIconsPack } from '@ui-kitten/eva-icons'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import { Provider as ThemeProvider } from '@berty-tech/components/theme'
import { FeatherIconsPack } from '@berty-tech/messenger-app/feather-icons'
import { CustomIconsPack } from '@berty-tech/messenger-app/custom-icons'
import { Provider as StyleProvider } from '@berty-tech/styles'

export const TestProvider: React.FC = ({ children }) => (
	<SafeAreaProvider
		initialMetrics={{
			frame: { x: 0, y: 0, width: 0, height: 0 },
			insets: { top: 0, left: 0, right: 0, bottom: 0 },
		}}
	>
		<StyleProvider>
			<IconRegistry icons={[EvaIconsPack, FeatherIconsPack, CustomIconsPack]} />
			<ThemeProvider>
				<NavigationContainer>{children}</NavigationContainer>
			</ThemeProvider>
		</StyleProvider>
	</SafeAreaProvider>
)

import _fixtures from './fixtures'
export const fixtures = _fixtures
