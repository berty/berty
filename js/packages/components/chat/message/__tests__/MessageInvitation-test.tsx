import React from 'react'
import renderer from 'react-test-renderer'

import beapi from '@berty-tech/api'
import { TestProvider } from '@berty-tech/testutil'

import { MessageInvitation } from '../MessageInvitation'

test('MessageInvitation renders correctly', () => {
	const props: React.ComponentProps<typeof MessageInvitation> = {
		message: {
			type: beapi.messenger.AppMessage.Type.TypeGroupInvitation,
			payload: {
				link: 'https://berty.tech/id#group/5QdUv6Fn3uvsxRCYb9hUn6UCe26QK7s7sqmFQGsQrNjj4TZmViFRyEwRr8dz3qMzGThz3QUTTj3NrAkg3WC6PqjXDvAp5UUrfzGqw8BxHiKye8QyRp46dJGAKQEjQfvkSx9GKm5QevrrDzsmfyUSkj6HqKVzPRZxr7X1FebqM7xxweY9K69tJoJC41an3fdbAr/name=some+group',
			},
		},
	}
	// FIXME: mock useContactRequest to get children in the snapshot
	const tree = renderer
		.create(
			<TestProvider>
				<MessageInvitation {...props} />
			</TestProvider>,
		)
		.toJSON()
	expect(tree).toMatchSnapshot()
})
