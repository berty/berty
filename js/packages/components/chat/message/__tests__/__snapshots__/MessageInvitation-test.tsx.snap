// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageInvitation renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "stretch",
            "flexDirection": "row",
            "justifyContent": "space-evenly",
          },
          Object {
            "paddingHorizontal": 16,
          },
          Object {
            "marginBottom": 9,
          },
          Object {
            "paddingTop": 2,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#F7F8FF",
            },
            Object {
              "padding": 16,
            },
            Object {
              "marginTop": 21,
            },
            Object {
              "width": 350,
            },
            Object {
              "borderRadius": 10,
            },
            Object {
              "shadowColor": "#000000",
              "shadowOffset": Object {
                "height": 2.5,
                "width": 0,
              },
              "shadowOpacity": 0.1,
              "shadowRadius": 4,
            },
            Object {},
          ]
        }
      >
        <View
          style={
            Object {
              "alignSelf": "center",
              "marginBottom": -28,
              "transform": Array [
                Object {
                  "translateY": -32.199999999999996,
                },
              ],
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                },
                Object {
                  "justifyContent": "center",
                },
                Object {
                  "width": 34,
                },
                Object {
                  "height": 34,
                },
                Object {
                  "borderRadius": 17,
                },
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#F7F8FF",
                  "borderWidth": 1,
                },
              ]
            }
          >
            <SvgMock
              height={22}
              style={
                Array [
                  Object {
                    "marginLeft": 4,
                  },
                ]
              }
              width={22}
            />
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "stretch",
                "flexDirection": "row",
                "justifyContent": "flex-start",
              },
              Object {
                "alignItems": "center",
              },
              Object {
                "justifyContent": "center",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 30,
                },
                Object {
                  "fontWeight": "bold",
                },
                Object {
                  "color": "#383B62",
                  "fontFamily": "Open Sans",
                },
              ]
            }
          >
            GROUP INVITATION
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "marginTop": 9,
              },
              Object {
                "alignItems": "center",
              },
              Object {
                "justifyContent": "center",
              },
            ]
          }
        >
          <View
            style={
              Object {
                "marginBottom": 9,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "position": "relative",
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 20,
                    "opacity": 1,
                    "shadowColor": "#000000",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": 0,
                    },
                    "shadowOpacity": 0.2,
                    "shadowRadius": 2,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "process": [Function],
                    }
                  }
                  style={
                    Object {
                      "borderRadius": 20,
                      "height": 40,
                      "width": 40,
                    }
                  }
                />
              </View>
              <View
                style={
                  Object {
                    "bottom": undefined,
                    "left": undefined,
                    "position": "absolute",
                    "right": -3,
                    "top": 3,
                  }
                }
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "backgroundColor": "#FFFFFF",
                      "borderRadius": 6.666666666666667,
                      "opacity": 1,
                      "shadowColor": "#000000",
                      "shadowOffset": Object {
                        "height": 1,
                        "width": 0,
                      },
                      "shadowOpacity": 0.2,
                      "shadowRadius": 2,
                    }
                  }
                >
                  <Image
                    source={
                      Object {
                        "process": [Function],
                      }
                    }
                    style={
                      Object {
                        "borderRadius": 6.666666666666667,
                        "height": 13.333333333333334,
                        "width": 13.333333333333334,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
          <Text
            style={
              Array [
                Object {
                  "fontSize": 26,
                },
                Object {
                  "fontWeight": "500",
                },
                Object {
                  "marginBottom": 9,
                },
                Object {
                  "color": "#383B62",
                  "fontFamily": "Open Sans",
                },
              ]
            }
          >
            
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "stretch",
                "flexDirection": "row",
                "justifyContent": "space-evenly",
              },
              Object {
                "justifyContent": "space-evenly",
              },
              Object {
                "alignItems": "center",
              },
              Object {
                "marginTop": 16,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "padding": 4,
                  },
                  Object {
                    "paddingVertical": 9,
                  },
                  Object {
                    "borderRadius": 4,
                  },
                  Object {
                    "width": 120,
                  },
                  Object {
                    "alignItems": "stretch",
                    "flexDirection": "row",
                    "justifyContent": "space-evenly",
                  },
                  Object {
                    "alignItems": "center",
                  },
                  Object {
                    "paddingRight": 9,
                  },
                  Object {
                    "backgroundColor": "#FFFFFF",
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "transform": Array [
                      Object {
                        "scale": 1,
                      },
                    ],
                  }
                }
              >
                <RNSVGSvgView
                  align="xMidYMid"
                  bbHeight={24}
                  bbWidth={24}
                  fill="#979797"
                  focusable={false}
                  height={24}
                  meetOrSlice={0}
                  minX={0}
                  minY={0}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "borderWidth": 0,
                      },
                      Array [
                        Object {
                          "opacity": 0.6,
                        },
                      ],
                      Object {
                        "flex": 0,
                        "height": 24,
                        "width": 24,
                      },
                    ]
                  }
                  vbHeight={24}
                  vbWidth={24}
                  width={24}
                >
                  <RNSVGGroup
                    fill={4288124823}
                    opacity={0.6}
                    propList={
                      Array [
                        "fill",
                      ]
                    }
                  >
                    <RNSVGGroup>
                      <RNSVGGroup>
                        <RNSVGRect
                          height="24"
                          matrix={
                            Array [
                              -1,
                              1.2246467991473532e-16,
                              -1.2246467991473532e-16,
                              -1,
                              24,
                              24,
                            ]
                          }
                          opacity={0}
                          width="24"
                          x={0}
                          y={0}
                        />
                        <RNSVGPath
                          d="M13.41 12l4.3-4.29a1 1 0 1 0-1.42-1.42L12 10.59l-4.29-4.3a1 1 0 0 0-1.42 1.42l4.3 4.29-4.3 4.29a1 1 0 0 0 0 1.42 1 1 0 0 0 1.42 0l4.29-4.3 4.29 4.3a1 1 0 0 0 1.42 0 1 1 0 0 0 0-1.42z"
                        />
                      </RNSVGGroup>
                    </RNSVGGroup>
                  </RNSVGGroup>
                </RNSVGSvgView>
              </View>
              <Text
                style={
                  Array [
                    Object {
                      "textAlign": "center",
                    },
                    Object {
                      "fontSize": 28,
                    },
                    Object {
                      "fontWeight": "bold",
                    },
                    Object {
                      "opacity": 0.6,
                    },
                    Object {
                      "color": "#979797",
                      "fontFamily": "Open Sans",
                    },
                  ]
                }
              >
                REFUSE
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "padding": 4,
                  },
                  Object {
                    "paddingVertical": 9,
                  },
                  Object {
                    "borderRadius": 4,
                  },
                  Object {
                    "width": 120,
                  },
                  Object {
                    "alignItems": "stretch",
                    "flexDirection": "row",
                    "justifyContent": "space-evenly",
                  },
                  Object {
                    "alignItems": "center",
                  },
                  Object {
                    "paddingRight": 9,
                  },
                  Object {
                    "backgroundColor": "#CED2FF",
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "transform": Array [
                      Object {
                        "scale": 1,
                      },
                    ],
                  }
                }
              >
                <RNSVGSvgView
                  align="xMidYMid"
                  bbHeight={24}
                  bbWidth={24}
                  fill="#525BEC"
                  focusable={false}
                  height={24}
                  meetOrSlice={0}
                  minX={0}
                  minY={0}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "borderWidth": 0,
                      },
                      Array [
                        Object {
                          "opacity": 0.6,
                        },
                      ],
                      Object {
                        "flex": 0,
                        "height": 24,
                        "width": 24,
                      },
                    ]
                  }
                  vbHeight={24}
                  vbWidth={24}
                  width={24}
                >
                  <RNSVGGroup
                    fill={4283587564}
                    opacity={0.6}
                    propList={
                      Array [
                        "fill",
                      ]
                    }
                  >
                    <RNSVGGroup>
                      <RNSVGGroup>
                        <RNSVGRect
                          height="24"
                          opacity={0}
                          width="24"
                          x={0}
                          y={0}
                        />
                        <RNSVGPath
                          d="M9.86 18a1 1 0 0 1-.73-.32l-4.86-5.17a1 1 0 1 1 1.46-1.37l4.12 4.39 8.41-9.2a1 1 0 1 1 1.48 1.34l-9.14 10a1 1 0 0 1-.73.33z"
                        />
                      </RNSVGGroup>
                    </RNSVGGroup>
                  </RNSVGGroup>
                </RNSVGSvgView>
              </View>
              <Text
                style={
                  Array [
                    Object {
                      "textAlign": "center",
                    },
                    Object {
                      "fontSize": 28,
                    },
                    Object {
                      "fontWeight": "bold",
                    },
                    Object {
                      "opacity": 0.6,
                    },
                    Object {
                      "color": "#525BEC",
                      "fontFamily": "Open Sans",
                    },
                  ]
                }
              >
                ACCEPT
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
